untrusted/urts/src/lib.rs

```diff
@@ -0,0 +1,197 @@
+            // modified buffer in another call to
+            // `sgx_create_enclave_from_buffer_ex()` then
+            // `SGX_ERROR_INVALID_ENCLAVE_ID` would be returned.
+            sgx_create_enclave_from_buffer_ex(
+                self.bytes.as_mut_ptr(),
+                self.bytes.len().try_into().unwrap(),
+                self.debug as c_int,
+                &mut enclave_id,
+                ptr::null_mut(),
+                0,
+                ptr::null_mut(),
+            )
+        };
+        match result {
+            sgx_status_t::SGX_SUCCESS => Ok(Enclave::new(enclave_id)),
+            error => Err(Error::SgxStatus(error)),
+        }
+    }
+}
+
+impl From<Vec<u8>> for EnclaveBuilder {
+    fn from(bytes: Vec<u8>) -> EnclaveBuilder {
+        EnclaveBuilder {
+            bytes,
+            debug: false,
+        }
+    }
+}
+
+impl Enclave {
+    fn new(id: sgx_enclave_id_t) -> Enclave {
+        Enclave { id }
+    }
+}
+
+impl Deref for Enclave {
+    type Target = sgx_enclave_id_t;
+    fn deref(&self) -> &Self::Target {
+        &self.id
+    }
+}
+
+impl Drop for Enclave {
+    /// Destroys the enclave through the SGX interface.
+    /// Any de-referenced value from [Enclave] is not valid after
+    /// dropping.
+    fn drop(&mut self) {
+        // Per the docs, this will only return SGX_SUCCESS or
+        // SGX_ERROR_INVALID_ENCLAVE_ID. The invalid ID error will only
+        // happen when the ID is invalid, the enclave hasn't been loaded,
+        // or the enclave has already been destroyed. Any of these cases
+        // don't afford corrective action, so ignore the return value
+        unsafe { sgx_destroy_enclave(self.id) };
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use test_enclave::{ecall_add_2, ENCLAVE};
+
+    #[test]
+    fn fail_to_create_enclave_with_non_existent_file() {
+        let builder = EnclaveBuilder::new(b"garbage bytes");
+        assert_eq!(
+            builder.create(),
+            Err(Error::SgxStatus(sgx_status_t::SGX_ERROR_INVALID_ENCLAVE))
+        );
+    }
+
+    #[test]
+    fn creating_enclave_succeeds() {
+        let builder = EnclaveBuilder::new(ENCLAVE);
+        assert!(builder.create().is_ok());
+    }
+
+    #[test]
+    fn calling_into_an_enclave_function_provides_valid_results() {
+        // Note: the `debug()` was added to ensure proper builder behavior of
+        // the `create()` method.  It could go away if another test has need
+        // of similar behavior.
+        let enclave = EnclaveBuilder::new(ENCLAVE).debug(true).create().unwrap();
+
+        let mut sum: c_int = 3;
+        let result = unsafe { ecall_add_2(*enclave, 3, &mut sum) };
+        assert_eq!(result, sgx_status_t::SGX_SUCCESS);
+        assert_eq!(sum, 3 + 2);
+    }
+
+    #[test]
+    fn create_enclave_builder_from_vector() {
+        let vector = ENCLAVE.to_vec();
+        assert!(EnclaveBuilder::from(vector).create().is_ok());
+    }
```

Comment:
Nit: Move instantiation tests closer together.
```suggestion

    #[test]
    fn create_enclave_builder_from_vector() {
        let vector = ENCLAVE.to_vec();
        assert!(EnclaveBuilder::from(vector).create().is_ok());
    }

    #[test]
    fn calling_into_an_enclave_function_provides_valid_results() {
        // Note: the `debug()` was added to ensure proper builder behavior of
        // the `create()` method.  It could go away if another test has need
        // of similar behavior.
        let enclave = EnclaveBuilder::new(ENCLAVE).debug(true).create().unwrap();

        let mut sum: c_int = 3;
        let result = unsafe { ecall_add_2(*enclave, 3, &mut sum) };
        assert_eq!(result, sgx_status_t::SGX_SUCCESS);
        assert_eq!(sum, 3 + 2);
    }
```